src/message.rs

```diff
@@ -62,3 +64,57 @@ impl fmt::Display for Message {
         )
     }
 }
+
+impl Message {
+    pub fn to_annotated_source(&self) -> String {
+        let source = fs::read_to_string(&self.filename).unwrap();
+        let error_lines = source
+            .lines()
+            .skip(self.location.row() - 1)
+            .take(self.end_location.row() - self.location.row() + 1)
+            .collect::<Vec<_>>();
+        let body = self.kind.body();
+        let code = self.kind.code().as_ref();
+        let rel_path = relativize_path(Path::new(&self.filename));
+        let source = error_lines.join("\n");
+        let range = (
+            self.location.column() - 1,
+            error_lines
+                .iter()
+                .enumerate()
+                .map(|(i, line)| {
+                    if i == error_lines.len() - 1 {
+                        self.end_location.column()
+                    } else {
+                        line.len() + 1
+                    }
+                })
+                .sum::<usize>()
+                - 1,
+        );
```

Comment:
Can we find a way to do all of the above through the cached `SourceCodeLocator` that we already create for each file? (Maybe it involves adding the body text to `message` or something?)

Reply:
> Maybe it involves adding the body text to message or something?

I'll try this.